diff -urN retroarch-v1.10.3/gfx/common/gl2_common.h ../build/retroarch-v1.10.3/gfx/common/gl2_common.h
--- a/retroarch-v1.10.3/gfx/common/gl2_common.h	2022-04-11 05:17:19.000000000 +0000
+++ b/retroarch-v1.10.3/gfx/common/gl2_common.h	2022-09-12 16:40:24.744036547 +0000
@@ -218,7 +218,7 @@
 
    struct scaler_ctx pbo_readback_scaler;
    struct video_viewport vp;                          /* int alignment */
-   math_matrix_4x4 mvp, mvp_no_rot;
+   math_matrix_4x4 mvp, mvp_no_rot, mvp_screen_rot;
    struct video_coords coords;                        /* ptr alignment */
    struct scaler_ctx scaler;
    video_info_t video_info;
diff -urN retroarch-v1.10.3/gfx/drivers/gl2.c ../build/retroarch-v1.10.3/gfx/drivers/gl2.c
--- a/retroarch-v1.10.3/gfx/drivers/gl2.c	2022-04-11 05:17:19.000000000 +0000
+++ b/retroarch-v1.10.3/gfx/drivers/gl2.c	2022-09-12 16:40:24.744036547 +0000
@@ -377,6 +377,10 @@
    matrix_4x4_ortho(gl->mvp_no_rot, ortho->left, ortho->right,
          ortho->bottom, ortho->top, ortho->znear, ortho->zfar);
 
+   matrix_4x4_rotate_z(rot, M_PI * 270 / 180.0f);
+   matrix_4x4_multiply(gl->mvp_screen_rot, rot, gl->mvp_no_rot);
+   gl->mvp_no_rot = gl->mvp_screen_rot;
+
    if (!allow_rotate)
    {
       gl->mvp = gl->mvp_no_rot;
@@ -468,7 +472,8 @@
       gl->vp.y *= 2;
 #endif
 
-   glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
+//   glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
+   glViewport(gl->vp.y, gl->vp.x, gl->vp.height, gl->vp.width);
    gl2_set_projection(gl, &default_ortho, allow_rotate);
 
    /* Set last backbuffer viewport. */
@@ -1882,7 +1887,7 @@
    glEnable(GL_BLEND);
 
    if (gl->overlay_full_screen)
-      glViewport(0, 0, width, height);
+      glViewport(0, 0, height, width);
 
    /* Ensure that we reset the attrib array. */
    gl->shader->use(gl, gl->shader_data,
@@ -1908,7 +1913,9 @@
    gl->coords.color     = gl->white_color_ptr;
    gl->coords.vertices  = 4;
    if (gl->overlay_full_screen)
-      glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
+      glViewport(gl->vp.y, gl->vp.x, gl->vp.height, gl->vp.width);
+
+//      glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
 }
 #endif
 
@@ -2465,9 +2472,10 @@
 
    if (gl->menu_texture_full_screen)
    {
-      glViewport(0, 0, width, height);
+      glViewport(0, 0, height, width);
       glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
-      glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
+//      glViewport(gl->vp.x, gl->vp.y, gl->vp.width, gl->vp.height);
+      glViewport(gl->vp.y, gl->vp.x, gl->vp.height, gl->vp.width);
    }
    else
       glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
@@ -2580,8 +2588,10 @@
       bounds = video_layout_screen(0);
 
       glViewport(
-         bounds->x, gl->video_height - bounds->y - bounds->h,
-         bounds->w, bounds->h
+//         bounds->x, gl->video_height - bounds->y - bounds->h,
+//         bounds->w, bounds->h
+         bounds->y, gl->video_width - bounds->x - bounds->w,
+         bounds->h, bounds->w
       );
    }
 }
diff -urN retroarch-v1.10.3/gfx/drivers_display/gfx_display_gl2.c ../build/retroarch-v1.10.3/gfx/drivers_display/gfx_display_gl2.c
--- a/retroarch-v1.10.3/gfx/drivers_display/gfx_display_gl2.c	2022-04-11 05:17:19.000000000 +0000
+++ b/retroarch-v1.10.3/gfx/drivers_display/gfx_display_gl2.c	2022-09-12 16:40:24.744036547 +0000
@@ -82,6 +82,8 @@
    1, 0
 };
 
+static float test = 0;
+
 static const float *gfx_display_gl2_get_default_vertices(void)
 {
    return &gl2_vertexes[0];
@@ -236,12 +238,14 @@
    if (!draw->coords->lut_tex_coord)
       draw->coords->lut_tex_coord = &gl2_tex_coords[0];
 
-   glViewport(draw->x, draw->y, draw->width, draw->height);
+   draw->y = draw->y;//video_height - draw->height - draw->y;
+
+   glViewport(draw->y, video_width - draw->width - draw->x, draw->height, draw->width);
    glBindTexture(GL_TEXTURE_2D, (GLuint)draw->texture);
 
    gl->shader->set_coords(gl->shader_data, draw->coords);
-   gl->shader->set_mvp(gl->shader_data,
-         draw->matrix_data ? (math_matrix_4x4*)draw->matrix_data
+   gl->shader->set_mvp(gl->shader_data, 
+	 draw->matrix_data ? (math_matrix_4x4*)draw->matrix_data
       : (math_matrix_4x4*)&gl->mvp_no_rot);
 
 
@@ -352,7 +356,7 @@
       int x, int y,
       unsigned width, unsigned height)
 {
-   glScissor(x, video_height - y - height, width, height);
+   glScissor(y, video_width - x - width, height, width);
    glEnable(GL_SCISSOR_TEST);
 #ifdef MALI_BUG
    /* TODO/FIXME: If video width/height changes between
@@ -372,7 +376,7 @@
       unsigned video_width,
       unsigned video_height)
 {
-   glScissor(0, 0, video_width, video_height);
+   glScissor(0, 0, video_height, video_width);
    glDisable(GL_SCISSOR_TEST);
 #ifdef MALI_BUG
    scissor_set_rectangle(0, video_width - 1, 0, video_height - 1, 0);
diff -urN retroarch-v1.10.3/gfx/drivers_shader/shader_glsl.c ../build/retroarch-v1.10.3/gfx/drivers_shader/shader_glsl.c
--- a/retroarch-v1.10.3/gfx/drivers_shader/shader_glsl.c	2022-04-11 05:17:19.000000000 +0000
+++ b/retroarch-v1.10.3/gfx/drivers_shader/shader_glsl.c	2022-09-12 16:40:24.744036547 +0000
@@ -94,6 +94,8 @@
    int frame_count;
    int frame_direction;
 
+   int rotation;
+
    int lut_texture[GFX_MAX_TEXTURES];
    unsigned frame_count_mod;
 
@@ -643,6 +645,8 @@
    uni->frame_count     = gl_glsl_get_uniform(glsl, prog, "FrameCount");
    uni->frame_direction = gl_glsl_get_uniform(glsl, prog, "FrameDirection");
 
+   uni->rotation        = gl_glsl_get_uniform(glsl, prog, "Rotation");
+
    for (i = 0; i < glsl->shader->luts; i++)
       uni->lut_texture[i] = glGetUniformLocation(prog, glsl->shader->lut[i].id);
 
@@ -1238,6 +1242,11 @@
          glUniform1i(uni->frame_direction, 1);
    }
 
+    if (uni->rotation >= 0)
+   {
+      glUniform1i(uni->rotation, retroarch_get_rotation());
+   }
+
    /* Set lookup textures. */
    for (i = 0; i < glsl->shader->luts; i++)
    {
