--- a/3rdparty/AudioCodecs/libopus/CMakeLists.txt	2024-09-23 21:48:19.000000000 +0000
+++ b/3rdparty/AudioCodecs/libopus/CMakelists.txt	2024-11-28 18:37:59.631971613 +0000
@@ -575,7 +575,7 @@
         celt/arm/arm_celt_map.c
     )
 
-    if(OPUS_ARM_ASM AND CMAKE_COMPILER_IS_GNUCC)
+    if(OPUS_ARM_ASM AND NOT OPUS_ARM64_ARCHITECTURE AND NOT OPUS_ARM_ARCHITECTURE)
         list(APPEND OPUS_SRC
             celt/arm/celt_pitch_xcorr_arm.s
         )
