diff --git a/gfx/gfx_display.c b/gfx/gfx_display.c
index 897326c321..baf57e9e5c 100644
--- a/gfx/gfx_display.c
+++ b/gfx/gfx_display.c
@@ -316,17 +316,6 @@ float gfx_display_get_dpi_scale(
        || (p_disp->menu_driver_id != last_menu_driver_id))
    {
       adjusted_scale            = scale * menu_scale_factor;
-#ifdef HAVE_OZONE
-      if (p_disp->menu_driver_id == MENU_DRIVER_ID_OZONE)
-      {
-         /* Ozone has a capped scale factor */
-         float new_width        = (float)width * 0.3333333f;
-         if (((float)OZONE_SIDEBAR_WIDTH * adjusted_scale)
-               > new_width)
-            adjusted_scale      = (new_width / (float)OZONE_SIDEBAR_WIDTH);
-      }
-#endif
-      adjusted_scale            = (adjusted_scale > 0.0001f) ? adjusted_scale : 1.0f;
       last_menu_scale_factor    = menu_scale_factor;
       last_menu_driver_id       = p_disp->menu_driver_id;
    }
