diff --git a/es-app/src/PlatformId.cpp b/es-app/src/PlatformId.cpp
index 8b751722..0e330ab5 100644
--- a/es-app/src/PlatformId.cpp
+++ b/es-app/src/PlatformId.cpp
@@ -198,6 +198,7 @@ namespace PlatformIds
 			
 		{ "pcarcade",				TEKNOPARROT },
 		{ "teknoparrot",			TEKNOPARROT },
+		{ "mame-advmame",                       MAME_ADVMAME },
 
 		{ "ignore",					PLATFORM_IGNORE },
 		{ "invalid",				PLATFORM_COUNT }
diff --git a/es-app/src/PlatformId.h b/es-app/src/PlatformId.h
index a085406a..9edf4437 100644
--- a/es-app/src/PlatformId.h
+++ b/es-app/src/PlatformId.h
@@ -175,6 +175,7 @@ namespace PlatformIds
 		LOWRESNX,
 
 		TEKNOPARROT,
+		MAME_ADVMAME,
 
 		PLATFORM_IGNORE, // do not allow scraping for this system
 		PLATFORM_COUNT
diff --git a/es-app/src/SystemData.cpp b/es-app/src/SystemData.cpp
index 13483972..b9beb94f 100644
--- a/es-app/src/SystemData.cpp
+++ b/es-app/src/SystemData.cpp
@@ -1672,7 +1672,7 @@ bool SystemData::isCheevosSupported()
 				"psx", "sg-1000", "sg1000", "coleco", "colecovision", "atari7800", "wonderswan", "pc88", "saturn", "3do", "apple2", "neogeo",
 				"arcade", "mame", "nds", "arcade", "megadrive-japan", "pcenginecd", "supergrafx", "supervision", "snes-msu1", "amstradcpc",
 				"dreamcast", "psp", "jaguar", "intellivision", "vectrex", "megaduck", "arduboy", "wasm4", "ps2", "gamecube", "wii", "channelf",
-				"o2em", "uzebox" };
+				"o2em", "uzebox", "mame-advmame" };
 
 			if (cheevosSystems.find(getName()) != cheevosSystems.cend())
 				mIsCheevosSupported = 1;
