diff --git a/es-app/src/PlatformId.cpp b/es-app/src/PlatformId.cpp
index 6178bc82..08aa72ae 100644
--- a/es-app/src/PlatformId.cpp
+++ b/es-app/src/PlatformId.cpp
@@ -197,6 +197,7 @@ namespace PlatformIds
 			
 		{ "pcarcade",				TEKNOPARROT },
 		{ "teknoparrot",			TEKNOPARROT },
+                { "mame-advmame",                       MAME_ADVMAME },
 
 		{ "ignore",					PLATFORM_IGNORE },
 		{ "invalid",				PLATFORM_COUNT }
diff --git a/es-app/src/PlatformId.h b/es-app/src/PlatformId.h
index 516608a5..1a181c2d 100644
--- a/es-app/src/PlatformId.h
+++ b/es-app/src/PlatformId.h
@@ -174,6 +174,7 @@ namespace PlatformIds
 		LOWRESNX,
 
 		TEKNOPARROT,
+		MAME_ADVMAME,
 
 		PLATFORM_IGNORE, // do not allow scraping for this system
 		PLATFORM_COUNT
diff --git a/es-app/src/SystemData.cpp b/es-app/src/SystemData.cpp
index 70439ce7..3252a1f4 100644
--- a/es-app/src/SystemData.cpp
+++ b/es-app/src/SystemData.cpp
@@ -1669,7 +1669,7 @@ bool SystemData::isCheevosSupported()
 				"atarilynx", "lynx", "ngp", "gamegear", "pokemini", "atari2600", "fbneo", "fbn", "virtualboy", "pcfx", "tg16", "famicom", "msx1",
 				"psx", "sg-1000", "sg1000", "coleco", "colecovision", "atari7800", "wonderswan", "pc88", "saturn", "3do", "apple2", "neogeo", "arcade", "mame",
 				"nds", "arcade", "megadrive-japan", "pcenginecd", "supergrafx", "supervision", "snes-msu1", "amstradcpc",
-				"dreamcast", "psp", "jaguar", "intellivision", "vectrex", "megaduck", "arduboy", "wasm4", "ps2", "gamecube", "wii" };
+				"dreamcast", "psp", "jaguar", "intellivision", "vectrex", "megaduck", "arduboy", "wasm4", "ps2", "mame-advmame", "gamecube", "wii" };
 
 			if (cheevosSystems.find(getName()) != cheevosSystems.cend())
 				mIsCheevosSupported = 1;
